### DISCUSSION TITLE
How to substitute values in ocelot.json file using Azure DevOps pipeline?

### DISCUSSION THREAD

**Discussion by saad569** [ThreeMammals/Ocelot#2010] [Q&A] [answered]
Hi i am using the ocelot json file in one of my API gateway, so i need to substitute values for the host/port for production environment which is different for DEV environment
How i can replace or substitute host/port value by using azure devops release pipeline.

**Comment by tnc1997** [comment, marked as answer]
Hi @saad569, one solution to differing hosts/ports could be to use an `ocelot.json` configuration file per environment, in a similar way to how the `appsettings.json` files work.

`ocelot.json` (default)
```json
  "ReRoutes": [
    {
      "DownstreamPathTemplate": "/downstreampathtemplate",
      "DownstreamScheme": "https",
      "DownstreamHostAndPorts": [
        {
          "Host": "host.azurewebsites.net",
          "Port": 443
        }
      ],
      "UpstreamHttpMethod": [
        "Get"
      ],
      "UpstreamPathTemplate": "/upstreampathtemplate"
    }
  ]
```

`ocelot.Development.json` (development)
```json
  "ReRoutes": [
    {
      "DownstreamPathTemplate": "/downstreampathtemplate",
      "DownstreamScheme": "https",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 5001
        }
      ],
      "UpstreamHttpMethod": [
        "Get"
      ],
      "UpstreamPathTemplate": "/upstreampathtemplate"
    }
  ]
```

`Program.cs`
```csharp
    public class Program
    {
        public static void Main(string[] args)
        {
            CreateHostBuilder(args).Build().Run();
        }

        public static IHostBuilder CreateHostBuilder(string[] args)
        {
            return Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.ConfigureAppConfiguration((context, builder) =>
                    {
                        builder
                            .SetBasePath(context.HostingEnvironment.ContentRootPath)
                            .AddJsonFile("appsettings.json", true, true)
                            .AddJsonFile($"appsettings.{context.HostingEnvironment.EnvironmentName}.json", true, true)
                            .AddJsonFile("ocelot.json", true, true)
                            .AddJsonFile($"ocelot.{context.HostingEnvironment.EnvironmentName}.json", true, true)
                            .AddEnvironmentVariables();
                    });

                    webBuilder.UseStartup<Startup>();
                });
        }
    }
```

`Startup.cs`
```csharp
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddOcelot();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
                app.UseDeveloperExceptionPage();
            else
                app.UseHsts();

            app.UseHttpsRedirection();

            app.UseOcelot().Wait();
        }
    }
```

Alternatively, if you prefer to apply transformations as part of a pipeline, then you could use the Azure App Service Deploy task. See [here](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/transforms-variable-substitution?view=azure-devops&tabs=Classic#json-variable-substitution) for more information regarding how to substitute variables.

**Reply by raman-m** [reply]
@tnc1997 Thomas, 
Thank you for the good and professional answer!

**Reply by raman-m** [reply]
Another solution, not using ASP.NET json-config files.
My suggestion is preparing some file template with placeholders.
Write a script or C# function to substitute placeholders with actual environment variable values.

**Comment by AndreasDamen** [comment]
> ```json
> host.azurewebsites.net
> ```

I tried your solution for using ocelot in azure. but somehow it is not working

**Reply by raman-m** [reply]
Hire me and I will help you!

**Reply by raman-m** [reply]
It was not working because of lack of your DevOps tasks!

**Comment by tnc1997** [comment]
> I tried your solution for using ocelot in azure. but somehow it is not working

Hi @AJCADamen, would you be able to elaborate on the issue that you are experiencing with this solution in more detail, for example by including your Ocelot configuration files and the build command that you are using to build the project?

**Comment by AndreasDamen** [comment]
> > I tried your solution for using ocelot in azure. but somehow it is not working
> 
> Hi @AJCADamen, would you be able to elaborate on the issue that you are experiencing with this solution in more detail, for example by including your Ocelot configuration files and the build command that you are using to build the project?

I set up everything correctly for deploying to azure. Host and port and also the baseurl. And the gateway works locally. But as soon as it is deployed it fails. 
Azure builds and deploys a linux docker container. This works fine considering i can reach the standard page after deployment.

![Screenshot_20210314-114519](https://user-images.githubusercontent.com/47353259/111065573-c93e0300-84ba-11eb-9d51-721b6a5d8c23.jpg)

**Reply by raman-m** [reply]
> Azure builds and deploys a linux docker container.

You must understand that Ocelot doesn't manage your Docker container!
You have to create Docker networks, care about environment variables, open access for file system etc... many many DevOps tasks

**Comment by tnc1997** [comment]
> But as soon as it is deployed it fails.

Is there any chance that you could expand on the failure that you are experiencing when it is deployed?

**Reply by raman-m** [reply]
In his case, there were no chances.
